core/external.rs

```diff
@@ -1,53 +1,258 @@
-use std::{marker::PhantomData, rc::Rc, cell::RefCell};
+use std::marker::PhantomData;
+use std::mem::ManuallyDrop;
 
 /// Define an external type.
 #[macro_export]
 macro_rules! external {
-  ($name:literal, $type:ident) => {
-      
+  ($type:ident, $name:literal) => {
+    impl $crate::external::Externalizable for $type {
+      fn external_marker() -> usize {
+        // Use the address of a static mut as a way to get around lack of const TypeId. Because it is mutable, the
+        // compiler cannot collapse multiple definitions into one.
+        static mut DEFINITION: $crate::external::ExternalDefinition =
+          $crate::external::ExternalDefinition::new($name);
+        // Wash the pointer through black_box so the compiler cannot see what we're going to do with it and needs
+        // to assume it will be used for valid purposes.
+        let ptr = std::hint::black_box(unsafe { &mut DEFINITION } as *mut _);
+        ptr as usize
```

Comment:
I don't understand this comment. This code isn't `const` either? Why is this preferrable over `TypeId::of`?

Reply:
Ack, I didn't update this comment when the design changed. TypeId isn't really a great fit for this because the implementation details change fairly often and we don't necessarily want to be testing a u128 for identity.